docs/madr/decisions/014-mesh-circuit-breaker-policy.md

```diff
@@ -0,0 +1,289 @@
+* Create single `MeshCircuitBreaker` with configuration divided
+  to `connectionThresholds` ([envoy's circuit breaker](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking))
+  and `outlierDetection` ([envoy's outlier detection](https://gstwww.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier))
+
+## Decision Outcome
+
+Chosen option: Create single `MeshCircuitBreaker` with configuration divided into two sections.
+
+## Positive Consequences
+
+* Having two separate policies, each with explicit configurations makes it easier to look for appropriate Envoy stats
+  and XDP configuration values.
+* By combining two abstractions, which exist in Envoy, and calling them both combined with a name of one of these is
+  confusing, especially when you are trying to debug some issue without deep understanding of internals of our service
+  mesh.
+* Default `MeshCircuitBreaker` policy is sufficient to replicate behaviour of current version of this policy - there
+  won't be a need to include a default `MeshOutlierDetector` policy
+
+## Negative Consequences
+
+* `MeshCircuitBreaker` will be a policy with very small amount of configuration available, which may introduce some
+  clutter (at some point there may be so many policies, that it will be much more difficult to understand all of them)
+
+## Solution
+
+By dividing the configuration section into two separate, logical sections: `connectionThresholds` and `outlierDetection`
+it's easier to differentiate underlying envoy features without the need to split policy into two separate ones.
+
+Introducing new - 2.0 policy is a great opportunity to introduce missing so far functionality of circuit breakers for
+inbound clusters, and with new policy matching it may be so simple to just allow to specify`from.targetRef` section
+
+Additional changes to the original `CircuitBreaker` policy consists of:
+
+- changing word `Errors` to `Failures` in detectors and adding `origin` suffix to the local failures
+  detector (`totalErrors`, `gatewayErrors` and `localErrors`becomes: `totalFailures`, `gatewayFailures`
+  and `localOriginFailures` respectively)
+- adding additional threshold-related property - `maxConnectionPools` for specifying the maximum number of connection
+  pools per cluster that Envoy will concurrently support at once
+
+### `connectionThresholds`
+
+Initially the plan was to call this part of the configuration `connectionPool`, but an Envoy allows to specify
+the `max_connection_pools` property for circuit breaking, I believe it wasn't the best choice.
+
+Kuma is currently not
+applying [priorities for routes](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-enum-config-core-v3-routingpriority),
+but to make it more future-proof `connectionThresholds` will be a collection with optional `priority` property
+defaulting to the `DEFAULT` value instead of a map.
```

Comment:
That's the hard choice, as it will be then much harder to change from:
```yaml
connectionThresholds:
  maxConnections: ...
  maxConnectionPools: ...
  maxPendingRequests: ...
  maxRetries: ...
  maxRequests: ...
```
to
```yaml
connectionThresholds:
- priority: DEFAULT|HIGH
  maxConnections: ...
  maxConnectionPools: ...
  maxPendingRequests: ...
  maxRetries: ...
  maxRequests: ...
```
as it will be a breaking change

Reply:
Maybe a list isn't required if you are able to configure only 2 values: `DEAFULT|HIGH`. Could we just introduce in the future `connectionThresholds.default|high` pointing to the same configuration options? When `connectionThresholds` config is for both and `connectionThresholds.default` might override the default.

Reply:
I like @lukidzi proposal. I'd keep the current one as a default and then add `connectionThresholds.high.max...` if needed

Reply:
Yeah I agree with Lukasz if it's a list just because there's 2 priorities it should be a map and if we add priority I see it something like:

```yaml
connectionThresholds:
  maxConnections: ...
  maxConnectionPools: ...
  maxPendingRequests: ...
  maxRetries: ...
  maxRequests: ...
  highPriorityOverrides:
     maxConnections: ... 
```

That way this is pure addition in the future.

@lobkovilya can confirm but as much as possible we should stay away from having lists

Reply:
sgtm

Reply:
addressed